crates/futures/src/polyfill.rs

```diff
@@ -0,0 +1,209 @@
+    let timeout = 0.0;
+    wait_async_with_timeout(indexed_array, index, value, timeout)
+}
+
+fn get_array_item(array: &JsValue, index: u32) -> JsValue {
+    Reflect::get(array, &JsValue::from(index))
+        .expect(&format!("Array should contain the index {}", index))
+}
+
+// Atomics.waitAsync always returns a promise. Throws standard errors
+// for parameter validation.  The promise is resolved with a string as from
+// Atomics.wait, or, in the case something went completely wrong, it is
+// rejected with an error string.
+pub fn wait_async_with_timeout(
+    indexed_array: Int32Array,
+    index: u32,
+    value: i32,
+    timeout: f64,
+) -> Result<Promise, JsValue> {
+    if !indexed_array.buffer().has_type::<SharedArrayBuffer>() {
+        console_log!("polyfill, not a SharedArrayBuffer");
+        return Err(Error::new("Indexed array must be created from SharedArrayBuffer").into());
+    }
+
+    // Optimization, avoid the helper thread in this common case.
+    if Atomics::load(&indexed_array, index)? != value {
+        console_log!("polyfill, not-equal");
+        return Ok(Promise::resolve(&JsString::from("not-equal")));
+    }
+
+    // General case, we must wait.
+
+    console_log!("polyfill, general case");
+
+    Ok(Promise::new(
+        &mut Box::new(move |resolve: Function, reject: Function| {
+            let helper = alloc_helper();
+            let helper_ref = helper.clone();
+
+            let onmessage_callback = Closure::wrap(Box::new(move |e: MessageEvent| {
+                // Free the helper early so that it can be reused if the resolution
+                // needs a helper.
+                free_helper(&helper_ref);
+                match String::from(
+                    get_array_item(&e.data(), 0)
+                        .as_string()
+                        .expect("data[0] should return a String"),
+                )
+                .as_str()
+                {
+                    "ok" => {
+                        resolve
+                            .call1(&JsValue::NULL, &get_array_item(&e.data(), 1))
+                            .expect("Should successfully call a resolve callback");
+                    }
+                    "error" => {
+                        // Note, rejection is not in the spec, it is an artifact of the polyfill.
+                        // The helper already printed an error to the console.
+                        reject
+                            .call1(&JsValue::NULL, &get_array_item(&e.data(), 1))
+                            .expect("Should successfully call a reject callback");
+                    }
+                    // it's not specified in the proposal yet
+                    _ => (),
+                }
+            }) as Box<dyn FnMut(MessageEvent)>);
+            helper
+                .borrow()
+                .set_onmessage(Some(onmessage_callback.as_ref().unchecked_ref()));
+
+            onmessage_callback.forget();
```

Comment:
I think we'll want to avoid `forget` here since it'll leak memory quickly, perhaps the `Closure::once` constructor could be used?

Reply:
`Closure::once` still produces `Error: closure invoked recursively or destroyed already` if I remove `forget`. May be `Closure::once_into_js` instead?